### PR TITLE
[App] Refactor plugins to be a standalone `LightningPlugin`

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Changes plugins to be standalone, similar to apps - rather than children of apps.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/app/core/flow.py [hunk 1]
```diff
@@ -741,22 +740,6 @@ def configure_api(self):
         """
         raise NotImplementedError
 
-    def configure_plugins(self) -> Optional[List[Dict[str, Plugin]]]:
```
Comment (tchaton) [inline src/lightning/app/core/flow.py:744]: Better not to have this hook for now.

File: src/lightning/app/core/plugin.py [hunk 2]
```diff
@@ -23,77 +25,41 @@
...
+        self.cluster_id = None
 
-    def run(self, name: str, entrypoint: str) -> None:
-        """Override with the logic to execute on the client side."""
+    def run(self, *args: str, **kwargs: str) -> None:
+        """Override with the logic to execute on the cloudspace."""
 
-    def run_app_command(self, command_name: str, config: Optional[BaseModel] = None) -> Dict[str, Any]:
-        """Run a command on the app associated with this plugin.
+    def run_job(self, name: str, app_entrypoint: str, env_vars: Optional[Dict[str, str]] = None) -> None:
```
Comment (tchaton) [inline src/lightning/app/core/plugin.py:47]:
```suggestion
    def run_job(self, name: str, app_entrypoint: str, env_vars: Optional[Dict[str, str]] = None, **kwargs) -> None:
```

More elements might be added like secrets, etc..

File: src/lightning/app/runners/cloud.py [hunk 3]
```diff
@@ -539,6 +550,23 @@ def _resolve_cloudspace_name(
...
         return cloudspace_name
 
+    def _resolve_run_name(
+        self,
+        name: str,
+        existing_instances: List[Externalv1LightningappInstance],
+    ) -> str:
+        """If there are existing instances with the same name - choose a randomised name."""
+        if len(existing_instances) > 0:
+            letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Comment (tchaton) [inline src/lightning/app/runners/cloud.py:560]: You can use python built-in: https://docs.python.org/3/library/string.html

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning/app/core/flow.py (hunk 1 above)
Comment (ethanwharris) [reply src/lightning/app/core/flow.py:744]: Yes, we can add back later if it makes sense

File: src/lightning/app/core/plugin.py (hunk 2 above)
Comment (ethanwharris) [reply src/lightning/app/core/plugin.py:47]: Yes, these can be added to the dispatch as needed, will leave for follow-up 😃

File: src/lightning/app/core/plugin.py [hunk 4]
```diff
@@ -103,54 +69,101 @@ def _run_plugin(run: _Run) -> None:
...
-                # This should never be possible but we check just in case
-                raise HTTPException(
-                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
-                    detail=f"The plugin {run.plugin_name} is an incorrect type.",
-                )
-    else:
-        raise HTTPException(
-            status_code=status.HTTP_400_BAD_REQUEST, detail="App ID must be specified unless `plugin_name='app'`."
-        )
+        # TODO: Return actions from the plugin here
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply src/lightning/app/core/plugin.py:165]: We will return actions directly for the UI to consume

File: src/lightning/app/runners/cloud.py (hunk 3 above)
Comment (ethanwharris) [reply src/lightning/app/runners/cloud.py]: Nice! Done 😃

File: src/lightning/__init__.py [hunk 5]
```diff
@@ -18,6 +18,7 @@
 from lightning.app import storage  # noqa: E402
 from lightning.app.core.app import LightningApp  # noqa: E402
 from lightning.app.core.flow import LightningFlow  # noqa: E402
+from lightning.app.core.plugin import LightningPlugin  # noqa: E402
```
Comment (Borda) [inline src/lightning/__init__.py:21]: Find is confusing with PL precisions plugins... :thinking:

File: src/lightning/app/core/plugin.py [hunk 6]
```diff
@@ -103,54 +69,101 @@ def _run_plugin(run: _Run) -> None:
...
-                cluster_id=run.cluster_id,
-            )
+            # Sometimes the URL gets encoded, so we parse it here
+            source_code_url = urlparse(run.source_code_url).geturl()
+
+            response = requests.get(source_code_url)
+
+            with open(download_path, "wb") as f:
+                f.write(response.content)
         except Exception as e:
```
Comment (Borda) [inline src/lightning/app/core/plugin.py:122]: Likely not all exceptions here...

File: src/lightning/app/runners/cloud.py [hunk 7]
```diff
@@ -232,10 +232,13 @@ def cloudspace_dispatch(
...
-        ignore_functions = self._resolve_open_ignore_functions()
-        repo = self._resolve_repo(root, ignore_functions)
-        cloudspace = self._resolve_cloudspace(project_id, cloudspace_id)
-        cluster_id = self._resolve_cluster_id(cluster_id, project_id, [cloudspace])
+        repo = self._resolve_repo(root)
+        self._resolve_cloudspace(project_id, cloudspace_id)
+        existing_instances = self._resolve_existing_run_instances(project_id, name)
+        name = self._resolve_run_name(
+            name,
+            existing_instances,
```
Comment (Borda) [inline src/lightning/app/runners/cloud.py:240]:
```suggestion
            existing_instances
```

File: src/lightning/app/runners/cloud.py [hunk 8]
```diff
@@ -521,6 +524,14 @@ def _resolve_existing_run_instance(
                     break
         return existing_cloudspace, existing_run_instance
 
+    def _resolve_existing_run_instances(self, project_id: str, name: str) -> List[Externalv1LightningappInstance]:
+        """Get all existing instances in the given project with the given name."""
+        run_instances = self.backend.client.lightningapp_instance_service_list_lightningapp_instances(
```
Comment (Borda) [inline src/lightning/app/runners/cloud.py:529]: What a horrible name, which does not tell me anything about what it does :D

File: src/lightning/app/runners/cloud.py [hunk 9]
```diff
@@ -529,16 +540,29 @@ def _resolve_cloudspace_name(
...
 
+    def _resolve_run_name(
+        self,
+        name: str,
+        existing_instances: List[Externalv1LightningappInstance],
+    ) -> str:
+        """If there are existing instances with the same name - choose a randomised name."""
+        if len(existing_instances) > 0:
+            name_exists = True
+            while name_exists:
```
Comment (Borda) [inline src/lightning/app/runners/cloud.py:559]: rather a foe loot with exit if you tried to many times

**Review by Borda — APPROVED** [review]

File: src/lightning/app/core/plugin.py (hunk 6 above)
Comment (ethanwharris) [reply src/lightning/app/core/plugin.py:122]: Needs to be all in this case as it's a web server, so any uncaught exceptions will give very unhelpful errors 😃

File: src/lightning/__init__.py (hunk 5 above)
Comment (ethanwharris) [reply src/lightning/__init__.py]: Removed top-level import, so it will be `L.app.LightningPlugin`

File: src/lightning/app/runners/cloud.py (hunk 8 above)
Comment (ethanwharris) [reply src/lightning/app/runners/cloud.py:526]: Changed to `_resolve_run_instances_by_name`

**Review by awaelchli — APPROVED** [review]

File: src/lightning/app/runners/cloud.py (hunk 9 above)
Comment (ethanwharris) [reply src/lightning/app/runners/cloud.py:556]: We will be refactoring this stuff quite a bit soon I think. Leaving for then

**Comment by awaelchli** [comment]
@ethanwharris Since this will probably merge soon, could I ask you to then cherry-pick that commit over into this branch: https://github.com/Lightning-AI/lightning/pull/16790 for an efficient releasing in the coming days. Thanks in advance!

**Review by justusschock — APPROVED** [review]

**Pull request merged**